src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -229,6 +234,8 @@ public void Initialize(IEventSource eventSource)
     /// <inheritdoc/>
     public void Shutdown()
     {
+        _cts.Cancel();
+        _refresher?.Join();
```

Comment:
In ideal case there'd as well be some signalization and `ThreadProc` would bail out immediately: https://github.com/dotnet/msbuild/blob/main/src/MSBuild/TerminalLogger/TerminalLogger.cs#L670 (e.g. `_cts.WaitHandle.WaitOne(1_000 / 30)`). And unless the `lock` wait is time limited as well - the `Join` here should probably have timeout ro prevent deadlock. Reasonably long wait - e.g. a second - is fine in both cases.

Reply:
That's a great point. The build currently spends ~16 ms on average waiting for the `Sleep` to finish.

Reply:
I agree that avoiding the extra waiting would be great, let's use the wait handle. As for the deadlock prevention, I here agree with Ladi that I do not like uncertainty of timeout in the Join. I do not yet see any opportunities for the deadlock here as well: the shutdown (and regular termination of TL) does not have a lock at all, since at this moment we are ok with possible race conditions that absence of lock may introduce. If the locked section is indefinitely long (because of the deadlock with something else for example), that is not a problem of introduced code, that is a problem we need to address for the rendering thread.

Reply:
I lack the knowledge of actual TL details - so I was just setting my eyes on seeing `Shutdown` waiting for a thread that can be waiting for a lock.
As soon as we know our locking is properly handled and hence `Shutdown` is guaranteed to finish - I'm not insisting on timeout.

I still feel bit strongly opinionated about replacing the `Sleep` with `WaitOne` though :-) - so waiting with sign-off for that one (or other contra-arguments)

Reply:
The WaitOne change is on its way. I will commit it soon :)